ksml-integration-tests/src/test/java/io/axual/ksml/integration/RepartitionIT.java

```diff
@@ -0,0 +1,218 @@
+ */
+@Slf4j
+@Testcontainers
+class RepartitionIT {
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    static final Network network = Network.newNetwork();
+
+    @Container
+    static final KafkaContainer kafka = new KafkaContainer("apache/kafka:4.0.0")
+            .withNetwork(network)
+            .withNetworkAliases("broker")
+            .withExposedPorts(9092, 9093);
+
+    @Container
+    static final KSMLContainer ksml = new KSMLContainer()
+            .withKsmlFiles("/docs-examples/reference/operations",
+                          "ksml-runner.yaml", "repartition-example-producer.yaml", "repartition-example-processor.yaml")
+            .withKafka(kafka)
+            .withTopics("user_activities", "repartitioned_activities")
+            .withPartitions(4)  // 4 partitions as per example
+            .dependsOn(kafka);
+
+    @Test
+    void testRepartitionWithCustomPartitioner() throws Exception {
+        // Wait for activities to be generated and repartitioned
+        log.info("Waiting for KSML to generate and repartition user activities...");
+        waitForActivityGeneration();
+
+        // Verify KSML is still running
+        assertThat(ksml.isRunning()).as("KSML should still be running").isTrue();
+
+        // Create consumer to read from repartitioned topic with partition info
+        final Properties consumerProps = createConsumerProperties();
+
+        Map<String, List<Integer>> userPartitions = new HashMap<>(); // Track which partitions each user goes to
+        List<ConsumerRecord<String, String>> allRecords = new ArrayList<>();
+
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(consumerProps)) {
+            consumer.subscribe(Collections.singletonList("repartitioned_activities"));
+            ConsumerRecords<String, String> records = KSMLRunnerTestUtil.pollWithRetry(consumer, Duration.ofSeconds(10));
+
+            assertThat(records).as("Should have repartitioned activities in repartitioned_activities topic").isNotEmpty();
+            log.info("Found {} repartitioned messages across {} partitions", records.count(), records.partitions().size());
+
+            // Collect all records with partition information
+            records.forEach(record -> {
+                allRecords.add(record);
+                String userId = record.key();
+                int partition = record.partition();
+
+                userPartitions.computeIfAbsent(userId, k -> new ArrayList<>()).add(partition);
+
+                log.info("Record: key={}, partition={}, value preview={}",
+                         userId, partition, record.value().substring(0, Math.min(100, record.value().length())));
+            });
+        }
+
+        // Validate we have records
+        assertThat(allRecords).as("Should have collected repartitioned records").isNotEmpty();
+
```

Comment:
Looks OK, the method is a bit long, actually there are 2 groups of assertions here; one about partitioning, and one about key transformation and value enrichment. Maybe in future see if we can split those out in separate methods (for readability).